{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "    Imports\n",
    "\"\"\"\n",
    "import numpy as np\n",
    "import operator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "    Dataset Loading\n",
    "    \n",
    "    Goal: Load the dataset and transform into the following format...\n",
    "    \n",
    "    Input: <str> -- filename of genotype data\n",
    "        - (39267,50) \n",
    "            * 39267 SNPs, 50 individuals\n",
    "        - SNP positions separated by row\n",
    "        - Each column is an individual\n",
    "        \n",
    "    Output: <np.array> -- (50,39267) \n",
    "        - Will use row index as unique identifier for person\n",
    "\"\"\"\n",
    "def loadGenotypes(filename):\n",
    "    with open(filename) as f:\n",
    "        raw_data = f.readlines()\n",
    "    data_out = np.array([[x.strip() for x in r.split(' ')] \\\n",
    "                         for r in raw_data]).T.astype(str)\n",
    "    return data_out\n",
    "    \n",
    "\"\"\"\n",
    "    Haplotype Phasing (Top Level)\n",
    "    \n",
    "    Goal: Top level function for haplotype phasing. Calls the \"E\" and \"M\"\n",
    "        step until we reach convergence (no more changes to haplotype\n",
    "        probabilities)\n",
    "        \n",
    "    Input: <np.array> -- (# individuals, genotype length)\n",
    "    \n",
    "    Output: {genotype: haplotype pair} -- mapping of genotype to its most\n",
    "        probable haplotype pair. \n",
    "\"\"\"\n",
    "def phaseHaplotypes(genotypes, window=100):\n",
    "    if window > len(genotypes[0]):\n",
    "        print(\"Please choose a window smaller than genotype length\")\n",
    "        return\n",
    "    \n",
    "    STOPPING_CRITERION = 0.01\n",
    "    \n",
    "    possible_haplos, geno_to_haplo = generateHaplotypes(genotypes, window)\n",
    "    \n",
    "    # Run EM until there's little average change in the probabilities\n",
    "    for i in range(100):\n",
    "        EStep(geno_to_haplo, possible_haplos)\n",
    "        avg_change = MStep(geno_to_haplo, possible_haplos)\n",
    "#         if avg_change < STOPPING_CRITERION:\n",
    "#             break\n",
    "    \n",
    "    decodeGenotypes(genotypes, geno_to_haplo, window)\n",
    "    \n",
    "\"\"\"\n",
    "    After running EM algorithm, decode the most probable haplotypes\n",
    "\"\"\"\n",
    "def decodeGenotypes(genotypes, geno_to_haplo, window):\n",
    "    decoded = []\n",
    "    for geno in genotypes:\n",
    "        geno = ''.join(geno)\n",
    "        h1 = \"\"\n",
    "        h2 = \"\"\n",
    "        for i in range(int(len(geno) / window)):\n",
    "            geno_slice = geno[i*window:(i+1)*window]\n",
    "            haplos = max(geno_to_haplo[geno_slice].items(), \\\n",
    "                         key=operator.itemgetter(1))[0]\n",
    "            h1 += haplos[0]\n",
    "            h2 += haplos[1]\n",
    "        \n",
    "        # Decode the rest of the genotype\n",
    "        for snp in geno[int(len(geno)/window)*window:]:\n",
    "            if snp == '0':\n",
    "                h1 += '0'\n",
    "                h2 += '0'\n",
    "            elif snp == '2':\n",
    "                h1 += '1'\n",
    "                h2 += '1'\n",
    "            else:\n",
    "                if np.random.binomial(1,0.5) == 0:\n",
    "                    h1 += '1'\n",
    "                    h2 += '0'\n",
    "                else:\n",
    "                    h1 += '0'\n",
    "                    h2 += '1'\n",
    "            \n",
    "        # Append h2 to end of h1\n",
    "        complete_haplo = h1 + h2\n",
    "        \n",
    "        # Add to decoded list, convert string to list\n",
    "        decoded.append([snp for snp in complete_haplo])\n",
    "        \n",
    "    # Make transpose, write to file\n",
    "    np.savetxt(\"haplo_soln.txt\", np.array(decoded).astype(str).T, \\\n",
    "              delimiter=' ', newline='\\r\\n', fmt=\"%s\")\n",
    "\"\"\"\n",
    "    Helper Function to Perform the E-Step\n",
    "\"\"\"\n",
    "def EStep(geno_to_haplo, possible_haplos):\n",
    "    for geno in geno_to_haplo:\n",
    "        normalized_geno_sum = 0\n",
    "        # Calculate haplotype probabiliies based on curr guesses\n",
    "        for haplo in geno_to_haplo[geno]:\n",
    "            haplo_prob = possible_haplos[haplo[0]] * \\\n",
    "                                possible_haplos[haplo[1]]\n",
    "            normalized_geno_sum += haplo_prob\n",
    "            geno_to_haplo[geno][haplo] = haplo_prob\n",
    "            \n",
    "        # Normalize probabilities per genotype\n",
    "        for haplo in geno_to_haplo[geno]:\n",
    "            geno_to_haplo[geno][haplo] /= normalized_geno_sum\n",
    "    \n",
    "\"\"\"\n",
    "    Helper Function to Perform the M-Step\n",
    "\"\"\"\n",
    "def MStep(geno_to_haplo, possible_haplos):\n",
    "    num_genos = len(geno_to_haplo)\n",
    "    avg_change = 0\n",
    "    for haplo in possible_haplos:\n",
    "        total_haplo_prob = 0\n",
    "        for geno in geno_to_haplo:\n",
    "            prob = getProbInGeno(geno_to_haplo[geno], haplo)\n",
    "            total_haplo_prob += prob\n",
    "        new_prob = total_haplo_prob / (2*num_genos)\n",
    "        avg_change += abs(new_prob - possible_haplos[haplo])\n",
    "        possible_haplos[haplo] = new_prob\n",
    "        \n",
    "    return avg_change / len(possible_haplos)\n",
    "        \n",
    "\"\"\"\n",
    "    M-Step helper to retrieve haplotype prob within a genotype\n",
    "\"\"\"\n",
    "def getProbInGeno(haplo_list, haplo):\n",
    "    for haplo_pair in haplo_list:\n",
    "        h1, h2 = haplo_pair\n",
    "        if h1 == haplo or h2 == haplo:\n",
    "            return haplo_list[haplo_pair]\n",
    "    return 0\n",
    "\n",
    "\"\"\"\n",
    "    Generate Haplotypes\n",
    "    \n",
    "    Goal: Help generate the list of haplotypes that can be produced from\n",
    "        the list of our genotypes\n",
    "        \n",
    "    Input: <np.array> -- (# individuals, genotype length)\n",
    "    \n",
    "    Output: \n",
    "        * {haplotype: haplotype probability} -- mapping of haplotype to\n",
    "            it's initial probability = 1/n\n",
    "        * {genotype: haplotypes} -- mapping of genotype to its associated\n",
    "            haplotypes\n",
    "\"\"\"\n",
    "def generateHaplotypes(genotypes, window):\n",
    "    possible_haplos = {}\n",
    "    geno_to_haplos = {}\n",
    "    visited_geno_slices = set()\n",
    "    for geno in genotypes:\n",
    "        # Convert to string for easier hashing\n",
    "        geno = ''.join(geno)\n",
    "        # Slice the geno into the window sizes\n",
    "        for i in range(int(len(geno) / window)):\n",
    "            geno_slice = geno[i*window:(i+1)*window]\n",
    "            \n",
    "            # Skip phasing for \n",
    "            if geno_slice in visited_geno_slices:\n",
    "                continue\n",
    "            else:\n",
    "                visited_geno_slices.add(geno_slice)\n",
    "                \n",
    "            # Generate map of genotype to its haplotype pairs\n",
    "            haplos = []\n",
    "            generateHaplotypesHelper(geno_slice, (\"\",\"\"), haplos)\n",
    "            # Remove mirrored tuples\n",
    "            haplos = list(set(tuple(sorted(l)) for l in haplos))\n",
    "            for h in haplos:\n",
    "                if geno_slice in geno_to_haplos:\n",
    "                    geno_to_haplos[geno_slice][h] = -1\n",
    "                else:\n",
    "                    geno_to_haplos[geno_slice] = {h: -1}\n",
    "                \n",
    "            \n",
    "            # Add to list of known haplotypes\n",
    "            for hap in haplos:\n",
    "                if hap[0] not in possible_haplos:\n",
    "                    possible_haplos[hap[0]] = 0\n",
    "                if hap[1] not in possible_haplos:\n",
    "                    possible_haplos[hap[1]] = 0\n",
    "                    \n",
    "    \n",
    "    num_haplos = len(possible_haplos)\n",
    "    for key in possible_haplos:\n",
    "        possible_haplos[key] = float(1/num_haplos)\n",
    "    return possible_haplos, geno_to_haplos\n",
    "        \n",
    "\"\"\"\n",
    "    Generate Haplotypes Helper\n",
    "    \n",
    "    Goal: Given a single genotype, recursively create a list of all the\n",
    "        possible haplotypes by looking at the genotype SNP by SNP. \n",
    "        Everytime we encounter a '1' in the genotype, make a recursive\n",
    "        branch.\n",
    "        \n",
    "    Input: <np.array> (genotype length,)\n",
    "    \n",
    "    Output: <np.array> (# haplotypes generated, haplotype length)\n",
    "\"\"\"\n",
    "def generateHaplotypesHelper(geno, curr_haplo, haplos):\n",
    "    if len(geno) == 0: \n",
    "        haplos.append(curr_haplo)\n",
    "        return\n",
    "    \n",
    "    # Consider the next genotype SNP\n",
    "    if geno[0] == '0':\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 ((curr_haplo[0] + '0'), \\\n",
    "                                 (curr_haplo[1] + '0')), \\\n",
    "                                 haplos)\n",
    "    elif geno[0] ==  '2':\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 (curr_haplo[0] + '1', \\\n",
    "                                 curr_haplo[1] + '1'), \\\n",
    "                                 haplos)\n",
    "    else:\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 (curr_haplo[0] + '0', \\\n",
    "                                 curr_haplo[1] + '1'), \\\n",
    "                                 haplos)\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 (curr_haplo[0] + '1', \\\n",
    "                                 curr_haplo[1] + '0'), \\\n",
    "                                 haplos)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "    MAIN CODE\n",
    "\"\"\"\n",
    "import time\n",
    "\n",
    "start = time.time()\n",
    "genotypes = loadGenotypes(\"example_data_1.txt\")\n",
    "phaseHaplotypes(genotypes, 40)\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
