{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "    Imports\n",
    "\"\"\"\n",
    "import numpy as np\n",
    "import operator\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "    Dataset Loading\n",
    "\n",
    "    Goal: Load the dataset and transform into the following format...\n",
    "\n",
    "    Input: <str> -- filename of genotype data\n",
    "        - (39267,50)\n",
    "            * 39267 SNPs, 50 individuals\n",
    "        - SNP positions separated by row\n",
    "        - Each column is an individual\n",
    "\n",
    "    Output: <np.array> -- (50,39267)\n",
    "        - Will use row index as unique identifier for person\n",
    "\"\"\"\n",
    "def loadGenotypes(filename):\n",
    "    with open(filename) as f:\n",
    "        raw_data = f.readlines()\n",
    "    data_out = np.array([[x.strip() for x in r.split(' ')] \\\n",
    "                         for r in raw_data]).T.astype(str)\n",
    "    return data_out\n",
    "\n",
    "\"\"\"\n",
    "    Haplotype Phasing (Top Level)\n",
    "\n",
    "    Goal: Top level function for haplotype phasing. Calls the \"E\" and \"M\"\n",
    "        step until we reach convergence (no more changes to haplotype\n",
    "        probabilities)\n",
    "\n",
    "    Input: <np.array> -- (# individuals, genotype length)\n",
    "\n",
    "    Output: {genotype: haplotype pair} -- mapping of genotype to its most\n",
    "        probable haplotype pair.\n",
    "\"\"\"\n",
    "def phaseHaplotypes(genotypes, window=50, overlap=3):\n",
    "    # TODO - magic number = 2x number of individuals\n",
    "    decoded = [\"\" for i in range(100)]\n",
    "    # First phase the initial chunk\n",
    "    curr_chunk = genotypes[:,:window]\n",
    "    curr_geno_haplos, curr_geno_to_haplo = \\\n",
    "        generateHaplotypes(curr_chunk, window)\n",
    "    for i in range(2):\n",
    "        EStep(curr_geno_to_haplo, curr_geno_haplos)\n",
    "        MStep(curr_geno_to_haplo, curr_geno_haplos)\n",
    "    cnt = 0\n",
    "    for curr_g in curr_chunk:\n",
    "        g_str = ''.join(curr_g)\n",
    "        haplos = max(curr_geno_to_haplo[g_str].items(),\\\n",
    "                        key=operator.itemgetter(1))[0]\n",
    "        decoded[cnt] += haplos[0]\n",
    "        decoded[cnt+1] += haplos[1]\n",
    "        cnt += 2\n",
    "        \n",
    "    # Then phase the overlap and next chunk\n",
    "    for i in range(int(len(genotypes[0]) / window)-1):\n",
    "        if i % 1000 == 0:\n",
    "            print(i)\n",
    "            print(np.array(decoded).shape)\n",
    "        overlap_chunk = genotypes[:,\\\n",
    "                            (i+1)*window-overlap:(i+1)*window+overlap]\n",
    "        overlap_geno_haplos, overlap_geno_to_haplo = \\\n",
    "            generateHaplotypes(overlap_chunk, 2*overlap)\n",
    "        next_chunk = genotypes[:,(i+1)*window:(i+2)*window]\n",
    "        next_geno_haplos, next_geno_to_haplo = \\\n",
    "            generateHaplotypes(next_chunk, window)\n",
    "\n",
    "        # Run EM on a few iterations for each map\n",
    "        for _ in range(2):\n",
    "            EStep(overlap_geno_to_haplo, overlap_geno_haplos)\n",
    "            MStep(overlap_geno_to_haplo, overlap_geno_haplos)\n",
    "            EStep(next_geno_to_haplo, next_geno_haplos)\n",
    "            MStep(next_geno_to_haplo, next_geno_haplos)\n",
    "        \n",
    "        # Decode this section while utilizing overlap\n",
    "        decodeGenotypes(genotypes[:,i*window:(i+2)*window], window, i,\\\n",
    "                        overlap, overlap_geno_to_haplo, \\\n",
    "                        next_geno_to_haplo, decoded)    \n",
    "    \n",
    "    print(\"Finished slicing\")\n",
    "    # Randomly decode the remainder xD\n",
    "    cnt = 0\n",
    "    for geno in genotypes:\n",
    "        geno = ''.join(geno)\n",
    "        h1 = \"\"\n",
    "        h2 = \"\"\n",
    "        for snp in geno[int(len(geno)/window)*window:]:\n",
    "            if snp == '0':\n",
    "                h1 += '0'\n",
    "                h2 += '0'\n",
    "            elif snp == '2':\n",
    "                h1 += '1'\n",
    "                h2 += '1'\n",
    "            else:\n",
    "                if np.random.binomial(1,0.5) == 0:\n",
    "                    h1 += '1'\n",
    "                    h2 += '0'\n",
    "                else:\n",
    "                    h1 += '0'\n",
    "                    h2 += '1'\n",
    "        decoded[cnt] += h1\n",
    "        decoded[cnt+1] += h2\n",
    "        cnt += 2\n",
    "    \n",
    "    decoded = [[s for s in string] for string in decoded]\n",
    "    \n",
    "    # Make transpose, write to file\n",
    "    np.savetxt(\"haplo_soln.txt\", np.array(decoded).astype(str).T, \\\n",
    "              delimiter=' ', newline='\\r\\n', fmt=\"%s\")\n",
    "\n",
    "def decodeGenotypes(geno_chunk, window, ind, overlap, \\\n",
    "                    overlap_geno_to_haplo, next_geno_to_haplo, decoded):\n",
    "    cnt = 0\n",
    "    for geno in geno_chunk:\n",
    "        geno_str = ''.join(geno)\n",
    "        \n",
    "        g1_h1 = decoded[cnt][ind*window:(ind+1)*window]\n",
    "        g1_h2 = decoded[cnt+1][ind*window:(ind+1)*window]\n",
    "        \n",
    "        overlap_str = geno_str[window-overlap:window+overlap]\n",
    "        o_h1, o_h2 = max(overlap_geno_to_haplo[overlap_str].items(),\\\n",
    "                            key=operator.itemgetter(1))[0]\n",
    "        \n",
    "        g2 = geno_str[window:]\n",
    "        g2_h1, g2_h2 = max(next_geno_to_haplo[g2].items(), \\\n",
    "                          key=operator.itemgetter(1))[0]\n",
    "        \n",
    "        perm1_h1 = g1_h1 + g2_h1\n",
    "        perm1_h2 = g1_h2 + g2_h2\n",
    "        perm2_h1 = g1_h1 + g2_h2\n",
    "        perm2_h2 = g1_h2 + g2_h1\n",
    "\n",
    "        # Find which ordering of g2's haplos is proper\n",
    "        if (o_h1 == perm1_h1[window-overlap:window+overlap] \\\n",
    "            and o_h2 == perm1_h2[window-overlap:window+overlap]) \\\n",
    "            or (o_h2 == perm1_h2[window-overlap:window+overlap] \\\n",
    "             and o_h1 == perm1_h2[window-overlap:window+overlap]):\n",
    "            decoded[cnt] += g2_h1\n",
    "            decoded[cnt+1] += g2_h2\n",
    "        else:\n",
    "            decoded[cnt] += g2_h2\n",
    "            decoded[cnt+1] += g2_h1\n",
    "        \n",
    "        cnt += 2    \n",
    "    \n",
    "\"\"\"\n",
    "    Helper Function to Perform the E-Step\n",
    "\"\"\"\n",
    "def EStep(geno_to_haplo, possible_haplos):\n",
    "    for geno in geno_to_haplo:\n",
    "        normalized_geno_sum = 0\n",
    "        # Calculate haplotype probabiliies based on curr guesses\n",
    "        for haplo in geno_to_haplo[geno]:\n",
    "            haplo_prob = possible_haplos[haplo[0]] * \\\n",
    "                                possible_haplos[haplo[1]]\n",
    "            normalized_geno_sum += haplo_prob\n",
    "            geno_to_haplo[geno][haplo] = haplo_prob\n",
    "\n",
    "        # Normalize probabilities per genotype\n",
    "        for haplo in geno_to_haplo[geno]:\n",
    "            geno_to_haplo[geno][haplo] /= normalized_geno_sum\n",
    "\n",
    "\"\"\"\n",
    "    Helper Function to Perform the M-Step\n",
    "\"\"\"\n",
    "def MStep(geno_to_haplo, possible_haplos):\n",
    "    num_genos = len(geno_to_haplo)\n",
    "    for haplo in possible_haplos:\n",
    "        total_haplo_prob = 0\n",
    "        for geno in geno_to_haplo:\n",
    "            prob = getProbInGeno(geno_to_haplo[geno], haplo)\n",
    "            total_haplo_prob += prob\n",
    "        new_prob = total_haplo_prob / (2*num_genos)\n",
    "        possible_haplos[haplo] = new_prob\n",
    "\n",
    "\"\"\"\n",
    "    M-Step helper to retrieve haplotype prob within a genotype\n",
    "\"\"\"\n",
    "def getProbInGeno(haplo_list, haplo):\n",
    "    for haplo_pair in haplo_list:\n",
    "        h1, h2 = haplo_pair\n",
    "        if h1 == haplo or h2 == haplo:\n",
    "            return haplo_list[haplo_pair]\n",
    "    return 0\n",
    "\n",
    "\"\"\"\n",
    "    Generate Haplotypes\n",
    "\n",
    "    Goal: Help generate the list of haplotypes that can be produced from\n",
    "        the list of our genotypes\n",
    "\n",
    "    Input: <np.array> -- (# individuals, genotype length)\n",
    "\n",
    "    Output:\n",
    "        * {haplotype: haplotype probability} -- mapping of haplotype to\n",
    "            it's initial probability = 1/n\n",
    "        * {genotype: haplotypes} -- mapping of genotype to its associated\n",
    "            haplotypes\n",
    "\"\"\"\n",
    "def generateHaplotypes(geno_chunk, window):\n",
    "    geno_haplos, geno_to_haplos = {}, {}\n",
    "    \n",
    "    # Create initial set of haplotypes from first geno chunk\n",
    "    sorted_geno_chunk = np.array(sorted(geno_chunk, \\\n",
    "                                key=lambda x: list(x).count('1')))    \n",
    "    init_geno = ''.join(sorted_geno_chunk[0])\n",
    "    haplos = []\n",
    "    generateHaplotypesHelper(init_geno, \\\n",
    "                             (\"\",\"\"), haplos)\n",
    "    haplos = list(set(tuple(sorted(l)) for l in haplos))\n",
    "    for h in haplos:\n",
    "        geno_to_haplos[init_geno] = {h: -1}\n",
    "        geno_haplos[h[0]] = 0\n",
    "        geno_haplos[h[1]] = 0\n",
    "        \n",
    "    for geno in sorted_geno_chunk:\n",
    "        geno_str = ''.join(geno)\n",
    "        # If the genotype is not phasable, add the closest possible\n",
    "        closest_phase = window\n",
    "        # Keep list of new haplos that partially phase\n",
    "        new_haplos = [] \n",
    "        # Potentially add closest haplo if no matches\n",
    "        close_haplo = None\n",
    "        for h in geno_haplos:\n",
    "            h1, h2, num_switches = phasable(h, geno_str)\n",
    "            if h1 > h2: # Keep ordering to help avoid duplicates\n",
    "                h1, h2 = h2, h1\n",
    "            if num_switches == 0:\n",
    "                new_haplos += [(h1, h2)]\n",
    "                closest_phase = -1 # No need to find closest phase\n",
    "            else:\n",
    "                if num_switches <= closest_phase:\n",
    "                    closest_phase = num_switches\n",
    "                    close_haplo = [(h1, h2)]\n",
    "        if closest_phase != -1:\n",
    "            if close_haplo is not None:\n",
    "                new_haplos += close_haplo\n",
    "            \n",
    "        # Empty out list of new haplos into haplo and geno_to_haplo maps\n",
    "        for new_h in new_haplos:\n",
    "            if new_h[0] not in geno_haplos:\n",
    "                geno_haplos[new_h[0]] = 0\n",
    "            if new_h[1] not in geno_haplos:\n",
    "                geno_haplos[new_h[1]] = 0\n",
    "            if geno_str in geno_to_haplos:\n",
    "                geno_to_haplos[geno_str][new_h] = -1\n",
    "            else:\n",
    "                geno_to_haplos[geno_str] = {new_h: -1}\n",
    "    \n",
    "    num_haplos = len(geno_haplos)\n",
    "    for key in geno_haplos:\n",
    "        geno_haplos[key] = float(1/num_haplos)\n",
    "    \n",
    "    return geno_haplos, geno_to_haplos\n",
    "        \n",
    "\"\"\"\n",
    "    Phasable checks if a haplotype phasing is valid for a given genotype\n",
    "    \n",
    "    Returns h1, the other half, and 0 if it's phasable\n",
    "    \n",
    "    Returns h1, the other half, and and min number of switches if it's \n",
    "        not phasable\n",
    "\"\"\"\n",
    "def phasable(h1, geno):\n",
    "    g_i = np.array([int(i) for i in geno])\n",
    "    h_i = np.array([int(i) for i in h1])\n",
    "    \n",
    "    num_switches = list(g_i-h_i).count(-1) + list(g_i-h_i).count(2)\n",
    "    \n",
    "    if num_switches > 0:\n",
    "        h_i = [0 if x == -1 else x for x in list(g_i-h_i)]\n",
    "        h_i = np.array([1 if x == 2 else x for x in h_i])\n",
    "        h1 = ''.join(h_i.astype(str))\n",
    "    h2 = ''.join((g_i - h_i).astype(str))\n",
    "    \n",
    "    return h1, h2, num_switches\n",
    "        \n",
    "\n",
    "\"\"\"\n",
    "    Generate Haplotypes Helper\n",
    "\n",
    "    Goal: Given a single genotype, recursively create a list of all the\n",
    "        possible haplotypes by looking at the genotype SNP by SNP.\n",
    "        Everytime we encounter a '1' in the genotype, make a recursive\n",
    "        branch.\n",
    "\n",
    "    Input: <np.array> (genotype length,)\n",
    "\n",
    "    Output: <np.array> (# haplotypes generated, haplotype length)\n",
    "\"\"\"\n",
    "def generateHaplotypesHelper(geno, curr_haplo, haplos):\n",
    "    if len(geno) == 0:\n",
    "        haplos.append(curr_haplo)\n",
    "        return\n",
    "\n",
    "    # Consider the next genotype SNP\n",
    "    if geno[0] == '0':\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 ((curr_haplo[0] + '0'), \\\n",
    "                                 (curr_haplo[1] + '0')), \\\n",
    "                                 haplos)\n",
    "    elif geno[0] ==  '2':\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 (curr_haplo[0] + '1', \\\n",
    "                                 curr_haplo[1] + '1'), \\\n",
    "                                 haplos)\n",
    "    else:\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 (curr_haplo[0] + '0', \\\n",
    "                                 curr_haplo[1] + '1'), \\\n",
    "                                 haplos)\n",
    "        generateHaplotypesHelper(geno[1:], \\\n",
    "                                 (curr_haplo[0] + '1', \\\n",
    "                                 curr_haplo[1] + '0'), \\\n",
    "                                 haplos)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39492\n",
      "0\n",
      "(100,)\n",
      "1000\n",
      "(100,)\n",
      "2000\n",
      "(100,)\n",
      "3000\n",
      "(100,)\n",
      "4000\n",
      "(100,)\n",
      "5000\n",
      "(100,)\n",
      "6000\n",
      "(100,)\n",
      "Finished window slicing\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "    MAIN CODE\n",
    "\"\"\"\n",
    "genotypes = loadGenotypes(\"example_data_1.txt\")\n",
    "phaseHaplotypes(genotypes, 6, 2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "cannot reshape array of size 3 into shape (3,4)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-07659d039129>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"asdf\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"asdf\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"asdf\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: cannot reshape array of size 3 into shape (3,4)"
     ]
    }
   ],
   "source": [
    "a = [\"asdf\", \"asdf\", \"asdf\"]\n",
    "b = np.array(a).reshape((3,4))\n",
    "print(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
